### DISCUSSION TITLE
1 Adder限制

### DISCUSSION THREAD

**Discussion by holyuming** [bol-edu/HLS-SOC-Discussions#95] [Q&A] [answered]
想請問1個adder的限制是指說只包含FIR計算的部分嗎?
還是說連counter...，address offset...計算這些都不能用到adder呢?

小弟不才，想不到如何不用counter卻知道什麼時候ap_done?

**Comment by JackeyUi** [comment]
同意這個問題，希望老師和助教能夠在workbook中定義更清楚的spec。
我希望能夠告訴我們 :
1. "register 限制" 具體是在 FIR 的計算過程中，還是整個FIR設計。因為 interface接收來的 read/write address和 write data，照理來講都應該要是用register去存，而不是直接用assign的方式，除此之外 data_length 的儲存和 data_num 的counting 也都需要register。
2. "一個加法器" 具體是在 FIR 的計算過程中，還是整個FIR設計。不管是address offset、data_length conting、bram address pointer等等，都需要counter，這些在synthesis 之後都會以adder的方式呈現。

**Comment by kevin861222** [comment]
同問，這個問題太重要了，另外想請問加法器、乘法器是否有bits限制？

**Comment by WeiChiaKung** [comment]
我的做法是
利用一個flag暫存器來確定是否是最後一筆資料，當ss_tlast為1且ss_tready也為1(亦即最後一筆輸入時):flag設為1
當flag為1，且sm_tvalid也被拉成1(亦即最後一筆輸出計算完時)，done為1
address offset 我則是用decode的方式解碼，而不是用減法
以上是我的作法，如有錯誤請指正。

**Comment by boledulab** [comment, marked as answer]
乘法器，加法器 (32-bit) 的限制僅在 FIR 算式中.
![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/98332019/1b7bd24e-9655-4523-94bf-84679d947040)

其他該需要用到的就用，bit 數應該隨設計的需要而定。

**Comment by boledulab** [comment]
事實上我很難想像同學們有人可以在整個 Lab3 中只用一個乘法器，一個加法器，如何做 Scheduling 就是個大問題，若真有人不惜時間代價，做了出來，一定給加分。
